### PR TITLE
Dict-like methods of ``Dataset``.

### LINKED ISSUE #76: dict-like API for Dataset
`Dataset.fields` is not very user friendly for iterating over values.

- Add methods like `keys`, `values`, etc.
- Should we also have __getitem__?

### PR DESCRIPTION
**Pull request by YooSunYoung**
Fixes #76 
During the [sprint](https://github.com/orgs/scipp/discussions/3267)

Dict-like methods:
- ``keys``
- ``values``
- ``items``
- ``__getitem__``
- ``__setitem__``
- ~``setdefault``~   We decided not to include this one.

Helper method to check if the key to the ``__(get/set)item__`` is valid:
- ``_validate_field_name``

## ``keys``/``values``/``items``
``keys`` return names of fields that belong to the current ``type`` of the ``Dataset`` object and other fields that only belong to the other ``type`` only if they are not ``None``.

``values`` and ``items`` are using ``keys``, hence using same set of fields.

## ``__getitem__``/``__setitem__``/~``setdefault``~
These methods check if the ``key``(``field_name``) is one of names of fields defined in the ``DatasetBase._FIELD_SPEC``.

If not valid, raise ``KeyError`` with custom error message.

But these methods don't check if they are valid fields of the current ``type``.

## Manual fields
Fields that are not defined in the ``DatasetBase._FIELD_SPEC`` are excluded for now. See #151 for more detail.

### CODE REVIEW

File: tests/dataset_test.py [hunk 1]
```diff
@@ -765,3 +765,136 @@ def test_derive_removes_attachments(initial, attachments):
...
+def invalid_field_example(my_type):
+    if my_type == DatasetType.DERIVED:
+        return "data_format", "sth_not_None"
+    elif my_type == DatasetType.RAW:
+        return "job_log_data", "sth_not_None"
+    else:
+        raise ValueError(my_type, " is not valid DatasetType.")
+
+
+def test_dataset_dict_like_keys_per_type(my_type):
```
Comment (jl-wynen) [inline tests/dataset_test.py:785]: Can you use the datasets strategy to generate input datasets? The strategy always makes valid datasets if you set `for_upload=True`.

**Comment by jl-wynen** [comment]
Can you add a release note?

File: src/scitacean/dataset.py [hunk 2]
```diff
@@ -507,6 +507,152 @@ def validate(self) -> None:
...
+
+        Raises
+        ------
+        :
+            :class:`KeyError` if ``field_name`` does not mach any names of fields.
+        """
+        self._validate_field_name(field_name)
+        setattr(self, field_name, field_value)
+
+    def setdefault(self, field_name: str, default_value: Any) -> Any:
```
[the remark this answers is not in the document]
Comment (YooSunYoung) [reply src/scitacean/dataset.py:625]: Yes that makes sense, if `None` actually means something. I will remove it.

File: src/scitacean/dataset.py [hunk 3]
```diff
@@ -507,6 +507,152 @@ def validate(self) -> None:
         """
         self.make_upload_model()
 
+    def keys(self) -> Iterable[str]:
+        """Dict-like keys(names of fields) method.
+
+        .. versionadded:: RELEASE_PLACEHOLDER
```
[the remark this answers is not in the document]
Comment (YooSunYoung) [reply src/scitacean/dataset.py]: image

It seems like it doesn't work if it is not above the sub-titles...?

[the remark this answers is not in the document]
Comment (jl-wynen) [reply src/scitacean/dataset.py:513]: This is an annoyance with restructured text. You need two empty lines to 'return' from the previous section. I just added this to the coding conventions: https://github.com/SciCatProject/scitacean/blob/main/docs/developer/coding-conventions.rst

**Comment by YooSunYoung** [comment]
@jl-wynen  I'm not authorized to merge this PR...!

**Pull request merged**